logic/tool/TypeResolver.java

```diff
@@ -144,355 +84,175 @@ public Conjunction resolveLabels(Conjunction conjunction) {
-                                constraint.asOwns().isKey());
-                    addNeighbours(newVar, constraint.asOwns().attribute());
-                    constraint.asOwns().overridden().ifPresent(typeVariable -> addNeighbours(newVar, typeVariable));
-                } else if (constraint.isPlays()) {
-                    newVar.plays(constraint.asPlays().relation().flatMap(var -> Optional.of(convert(var))).orElse(null),
-                                 convert(constraint.asPlays().role()),
-                                 constraint.asPlays().overridden().flatMap(var -> Optional.of(convert(var))).orElse(null));
-                    addNeighbours(newVar, constraint.asPlays().role());
-                    constraint.asPlays().overridden().ifPresent(typeVariable -> addNeighbours(newVar, typeVariable));
-                } else if (constraint.isRelates()) {
-                    newVar.relates(convert(constraint.asRelates().role()),
-                                   constraint.asRelates().overridden().flatMap(var -> Optional.of(convert(var))).orElse(null));
-                    addNeighbours(newVar, constraint.asRelates().role());
-                    constraint.asRelates().overridden().ifPresent(typeVariable -> addNeighbours(newVar, typeVariable));
-                } else if (constraint.isIs()) {
-                    newVar.is(convert(constraint.asIs().variable()));
-                } else throw GraknException.of(ILLEGAL_STATE);
+            Set<ValueType> valueTypes = findComparableValueTypes(constraint);
+            if (valueTypeRegister.get(owner.id()).isEmpty()) {
+                valueTypes.forEach(valueType -> resolverTraversal.valueType(owner.id(), valueType));
+                valueTypeRegister.put(owner.id(), valueTypes);
+            } else if (!valueTypeRegister.get(owner.id()).containsAll(valueTypes)) {
+                throw GraknException.of(SCHEMATICALLY_UNSATISFIABLE_CONJUNCTION, conjunction);
             }
-            return newVar;
+            subAttribute(owner.id());
         }
 
-    }
-
-    private static class Resolvers {
-
-        private final Map<Identifier, TypeVariable> typeResolvers;
-        private final Map<RelationConstraint.RolePlayer, TypeVariable> roleTypeResolvers;
-        // TODO: Why is there no checkers and getters for 'roleTypeResolvers'?
-        // TODO: do we still need 'roleTypeResolvers' given that we no longer put resolveRoleTypes on RelationConstraints?
-
-        private int sysVarCounter;
-
-        Resolvers() {
-            this.sysVarCounter = 0;
-            this.typeResolvers = new HashMap<>();
-            this.roleTypeResolvers = new HashMap<>();
-        }
-
-        TypeVariable register(RelationConstraint.RolePlayer rolePlayer) {
-            return roleTypeResolvers.computeIfAbsent(rolePlayer, rp -> new TypeVariable(newSystemVariable()));
+        private Set<ValueType> findComparableValueTypes(ValueConstraint<?> constraint) {
```

Comment:
Turns out convertValue is tricky. Consider

```
$x = 2;
$y = $x;
```

we'd like to convert this on the resolver traversal to be:
```
$x value long;
$y value $x
```

But this isn't possible, since `valueTypeConstraint` cannot have a `Variable` as `value` (unlike `ValueConstraint`, which can).

Therefore, we have to convert it directly to:
```
$x value long;
$y value;
```

We do this by converting any `variable` we see in `value`, and then returning that as the valueType. This is why we need the `valueTypeRegister`.

But there's another problem! `$x` also can be equal to 2.0. So we need to return the valueType *comparable* not *equal* to the value type.